frame/nis/src/lib.rs

```diff
@@ -704,35 +760,176 @@ pub mod pallet {
+		/// - `index`: The index of the receipt.
+		#[pallet::call_index(4)]
+		#[pallet::weight(T::WeightInfo::thaw_communal())]
+		pub fn thaw_communal(
+			origin: OriginFor<T>,
+			#[pallet::compact] index: ReceiptIndex,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Look for `index`
+			let receipt: ReceiptRecordOf<T> =
+				Receipts::<T>::get(index).ok_or(Error::<T>::UnknownReceipt)?;
+			// If found, check it is actually communal.
+			ensure!(receipt.owner.is_none(), Error::<T>::NotOwner);
+			let now = frame_system::Pallet::<T>::block_number();
+			ensure!(now >= receipt.expiry, Error::<T>::NotExpired);
+
+			let mut summary: SummaryRecordOf<T> = Summary::<T>::get();
+
+			let (throttle, throttle_period) = T::ThawThrottle::get();
+			if now.saturating_sub(summary.last_period) >= throttle_period {
+				summary.thawed = Zero::zero();
+				summary.last_period = now;
+			}
+			summary.thawed.saturating_accrue(receipt.proportion);
+			ensure!(summary.thawed <= throttle, Error::<T>::Throttled);
+
+			T::Counterpart::burn_from(&who, T::CounterpartAmount::convert(receipt.proportion))?;
+
+			// Multiply the proportion it is by the total issued.
+			let our_account = Self::account_id();
+			let effective_issuance = Self::issuance_with(&our_account, &summary).effective;
+			let amount = receipt.proportion * effective_issuance;
+
+			summary.proportion_owed.saturating_reduce(receipt.proportion);
+
+			// Try to transfer amount owed from pot to receipt owner.
+			T::Currency::transfer(&our_account, &who, amount, AllowDeath)
+				.map_err(|_| Error::<T>::Unfunded)?;
+
+			Receipts::<T>::remove(index);
+			Summary::<T>::put(&summary);
+
+			let e =
+				Event::Thawed { index, who, amount, proportion: receipt.proportion, dropped: true };
+			Self::deposit_event(e);
+
+			Ok(())
+		}
+
+		/// Make a private receipt communal and create fungible counterparts for its owner.
+		#[pallet::call_index(5)]
+		#[pallet::weight(T::WeightInfo::communify())]
+		pub fn communify(
+			origin: OriginFor<T>,
+			#[pallet::compact] index: ReceiptIndex,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Look for `index`
+			let mut receipt: ReceiptRecordOf<T> =
+				Receipts::<T>::get(index).ok_or(Error::<T>::UnknownReceipt)?;
+
+			// Check it's not already communal and make it so.
+			let (owner, on_hold) = receipt.owner.take().ok_or(Error::<T>::AlreadyCommunal)?;
+
+			// If found, check the owner is `who`.
+			ensure!(owner == who, Error::<T>::NotOwner);
+
+			// Unreserve and transfer the funds to the pot.
+			T::Currency::unreserve_named(&T::ReserveId::get(), &who, on_hold);
+			// Transfer `excess` to the pot if we have now fully compensated for the receipt.
+			T::Currency::transfer(&who, &Self::account_id(), on_hold, AllowDeath)
+				.map_err(|_| Error::<T>::Unfunded)?; //< Requires transactional storage TODO: OK?
+									 // TODO: ^^^ The above should be done in a single operation `transfer_on_hold`.
```

Comment:
Calls are all wrapped in a transactional layer now, so the storage changes of the whole call revert on error return.  
There is also `with_storage_layer` to wrap a closure in a new layer, but it would make no difference here.